concent_api/core/tests/test_upload_finished_task.py

```diff
@@ -0,0 +1,82 @@
+            task_id='1',
+            subtask_id='8',
+            provider_public_key=self.PROVIDER_PUBLIC_KEY,
+            requestor_public_key=self.REQUESTOR_PUBLIC_KEY,
+            state=Subtask.SubtaskState.VERIFICATION_FILE_TRANSFER,
+            next_deadline=get_current_utc_timestamp() + settings.CONCENT_MESSAGING_TIME,
+            task_to_compute=self._get_deserialized_task_to_compute(
+                task_id='1',
+                subtask_id='8',
+            ),
+            report_computed_task=self._get_deserialized_report_computed_task(
+                subtask_id='8',
+            )
+        )
+
+    def test_that_scheduling_task_for_subtask_with_accepted_or_failed_or_additional_verification_state_should_log_warning_and_finish_task(self):
+        for state in [
+            Subtask.SubtaskState.ACCEPTED,
+            Subtask.SubtaskState.FAILED,
+            Subtask.SubtaskState.ADDITIONAL_VERIFICATION,
+        ]:
+            Subtask.objects.filter(subtask_id=self.subtask.subtask_id).update(state=state.name)  # pylint: disable=no-member
+            self.subtask.refresh_from_db()
+
+            with mock.patch('core.tasks.logging.warning') as logging_warning_mock:
+                upload_finished(self.subtask.subtask_id)  # pylint: disable=no-value-for-parameter
+
+            logging_warning_mock.assert_called_once_with(
+                f'Subtask with ID {self.subtask.subtask_id} is expected to be in `VERIFICATION_FILE_TRANSFER` state, but was in {self.subtask.state}.'
+            )
+
+    def test_that_scheduling_task_for_subtask_with_unexpected_state_should_log_warning_and_finish_task(self):
+        Subtask.objects.filter(subtask_id=self.subtask.subtask_id).update(state=Subtask.SubtaskState.REPORTED.name)  # pylint: disable=no-member
+        self.subtask.refresh_from_db()
+
+        with mock.patch('core.tasks.logging.error') as logging_error_mock:
+            upload_finished(self.subtask.subtask_id)  # pylint: disable=no-value-for-parameter
+
+        logging_error_mock.assert_called_once_with(
+            f'Subtask with ID {self.subtask.subtask_id} is expected to be in `VERIFICATION_FILE_TRANSFER` state, but was in {self.subtask.state}.'
+        )
+
+    def test_that_scheduling_task_for_subtask_before_deadline_should_change_subtask_state_and_schedule_upload_acknowledged_task(self):
+        with freeze_time(parse_timestamp_to_utc_datetime(self.subtask.next_deadline.timestamp() - 1)):
+            with mock.patch('core.tasks.upload_acknowledged.delay') as upload_acknowledged_delay_mock:
+                upload_finished(self.subtask.subtask_id)  # pylint: disable=no-value-for-parameter
+
+        self.subtask.refresh_from_db()
+        self.assertEqual(self.subtask.state_enum, Subtask.SubtaskState.ADDITIONAL_VERIFICATION)
+        upload_acknowledged_delay_mock.assert_called_once_with(self.subtask.subtask_id)
+
+    def test_that_scheduling_task_for_subtask_after_deadline_should_process_timeout(self):
```

Comment:
shouldn't this check payment as well?

Reply:
`mock` backend is used. I suggest patching payment function, so call to to it can be asserted.

Reply:
done

Reply:
I thought we agreed that when the mock backend is used we don't have to mock those calls. But actually I prefer that we do this. So lets agree on one way and stick to it. And make it as our good practice. And add it where it is missing.